src/main/java/com/aws/greengrass/security/SecurityService.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.aws.greengrass.security;
+
+import com.aws.greengrass.config.CaseInsensitiveString;
+import com.aws.greengrass.logging.api.Logger;
+import com.aws.greengrass.logging.impl.LogManager;
+import com.aws.greengrass.security.exceptions.KeyLoadingException;
+import com.aws.greengrass.security.exceptions.ServiceProviderConflictException;
+import com.aws.greengrass.security.exceptions.ServiceUnavailableException;
+import lombok.AccessLevel;
+import lombok.Getter;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import javax.net.ssl.KeyManager;
+
+public class SecurityService {
+    private static final Logger logger = LogManager.getLogger(SecurityService.class);
+    private static final String KEY_TYPE = "keyType";
+    private static final String KEY_URI = "keyUri";
+
+    @Getter(AccessLevel.PACKAGE)
+    private final ConcurrentMap<CaseInsensitiveString, CryptoKeySpi> cryptoKeyProviderMap;
+
+    public SecurityService() {
+        this.cryptoKeyProviderMap = new ConcurrentHashMap<>();
+    }
+
+    /**
+     * Register crypto key provider for the key type.
+     *
+     * @param keyProvider Crypto key provider
+     * @throws ServiceProviderConflictException if key type is already registered
+     */
+    public void registerCryptoKeyProvider(CryptoKeySpi keyProvider) throws ServiceProviderConflictException {
+        CaseInsensitiveString keyType = new CaseInsensitiveString(keyProvider.supportedKeyType());
+        logger.atInfo().kv(KEY_TYPE, keyType).log("Register crypto key service provider");
+        CryptoKeySpi provider = cryptoKeyProviderMap.computeIfAbsent(keyType, k -> keyProvider);
+        if (!provider.equals(keyProvider)) {
+            logger.atError().kv(KEY_TYPE, keyType)
+                    .log("Crypto key service provider for the key type is already registered");
+            throw new ServiceProviderConflictException(String.format("Key type %s provider is registered", keyType));
+        }
+    }
+
+    /**
+     * Deregister crypto key provide for the key type.
+     *
+     * @param keyProvider Crypto key provider
+     */
+    public void deregisterCryptoKeyProvider(CryptoKeySpi keyProvider) {
+        CaseInsensitiveString keyType = new CaseInsensitiveString(keyProvider.supportedKeyType());
+        boolean removed = cryptoKeyProviderMap.remove(keyType, keyProvider);
+        if (!removed) {
+            logger.atInfo().kv(KEY_TYPE, keyType).log("Crypto key service provider is either removed or not the same"
+                    + " provider");
```

Comment:
To avoid confusion, this probably can be simplified to : _Crypto key service provider is either already removed or is unregistered._

Reply:
removed equals to unregistered in this case. There is case that caller tried to remove a provider with same key type but different instance. In such case, the existing instance shouldn't be removed.

Reply:
We can think of it like this:
`register(keyType1, provider1);`    <--- here `CryptoKeyProviderMap` will have entry `{keyType1, provider1}` . This Concurrent map will always hold only _one_ entry for a CaseInsensitive keyType string.

`deregister(keyType1, provider2);`  <---- Note that its the same keyType1 but different provider which was not registered before. `CryptoKeyProviderMap` will not have `provider2` entry value for `keyType1`, because provider2 was not registered previously for keyType1. i.e. provider2 is an _unregistered_ provider.

Reply:
Make sense. I will change wording in the next revision. Thanks!